### PR TITLE
Bug 1796522: Show multiple resources in one stream

### PR DESCRIPTION
**Pull request by zherman0**
We now have the ability the select multiple event streams at once. Instead of showing each stream in its own section, I have combined the events to show together which is the preference for users.
Peek 2020-02-12-events

### CODE REVIEW

**Comment by zherman0** [comment]
/assign @benjaminapetersen 
@benjaminapetersen  - this is ready for review as well

File: frontend/public/components/events.jsx [hunk 1]
```diff
@@ -64,14 +64,23 @@ const kindFilter = (reference, { involvedObject }) => {
...
   }
-
-  if (isGroupVersionKind(reference)) {
-    const group = apiGroupForReference(reference);
-    const kind = kindForReference(reference);
-    const { group: eventGroup } = groupVersionFor(involvedObject.apiVersion);
-    return involvedObject.kind === kind && eventGroup === group;
-  }
-  return involvedObject.kind === reference;
+  let filterKind = false;
```
Comment (benjaminapetersen) [inline frontend/public/components/events.jsx]: I would change the impl here from `let filterKind =false` to a `const` and use `_.filter` or a better util than `_.map`.

Comment (benjaminapetersen) [reply frontend/public/components/events.jsx]: I'm not super worried about the form tbh, this is reasonable as a bug fix.  Perhaps we can throw a unit test up for it quick?

File: frontend/public/components/events.jsx [hunk 2]
```diff
@@ -64,14 +64,23 @@ const kindFilter = (reference, { involvedObject }) => {
...
+  const kinds = reference.indexOf(',') > 0 ? reference.split(',') : [reference];
```
Comment (spadgett) [inline frontend/public/components/events.jsx]: This is the same as

```suggestion
  const kinds = reference.split(',');
```

File: frontend/public/components/events.jsx [hunk 3]
```diff
@@ -64,14 +64,23 @@ const kindFilter = (reference, { involvedObject }) => {
...
+  kinds.map((ref) => {
```
Comment (spadgett) [inline frontend/public/components/events.jsx]: `Array.some()` is better here

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some

```ts
return kinds.some((ref: string) => {
  if (isGroupVersionKind(ref)) {
    const group = apiGroupForReference(ref);
    const kind = kindForReference(ref);
    const { group: eventGroup } = groupVersionFor(involvedObject.apiVersion);
    return involvedObject.kind === kind && eventGroup === group);
  }

  return involvedObject.kind === ref;
};
```

File: frontend/public/components/events.jsx [hunk 4]
```diff
@@ -196,24 +205,24 @@ export class EventsList extends React.Component {
         </div>,
       );
     } else {
-      selected.forEach((kind) => {
-        events.push(
-          <div key={kind}>
-            <span>
+      events.push(
```
Comment (spadgett) [inline frontend/public/components/events.jsx]: This shouldn't need to be an array anymore, right?

We should be able to go back to what was essentially the logic in 4.3, correct? We just need to update the filter method. I think this can be greatly simplified now.

File: frontend/integration-tests/tests/crud.scenario.ts [hunk 5]
```diff
@@ -148,6 +148,16 @@ describe('Kubernetes resource CRUD operations', () => {
         expect(crudView.resourceTitle.getText()).toEqual(name);
       });
 
+      it('event view displays created resource instance', async () => {
```
Comment (spadgett) [inline frontend/integration-tests/tests/crud.scenario.ts]: Hey @zherman0 I'm confused by this test change. This doesn't look right to me. Why is this change needed?

Comment (zherman0) [reply frontend/integration-tests/tests/crud.scenario.ts]: I was trying to add an event page specific test, but that proves to be the wrong place.  I will clean up that file shortly.

Comment (spadgett) [reply frontend/integration-tests/tests/crud.scenario.ts]: OK, I see. I didn't realize we had this test gap. Thanks for adding something 👍 

It should be outside this loop or it will re-run the tests for every resource.

Comment (zherman0) [reply frontend/integration-tests/tests/crud.scenario.ts]: I purposely was trying to leave it in that loop, but it turns out that not all the resources show on the events page, which is why my test was failing.  I have created a specific block for it to run in and I will just test a few resources that I know will show up.

File: frontend/integration-tests/tests/crud.scenario.ts [hunk 6]
```diff
@@ -263,6 +263,55 @@ describe('Kubernetes resource CRUD operations', () => {
...
+      await yamlView.setEditorContent(safeDump(newContent));
+      leakedResources.add(JSON.stringify({ name, plural: 'pods', testName }));
+      await yamlView.saveButton.click();
+      expect(crudView.errorMessage.isPresent()).toBe(false);
+    });
+
+    it('event view displays created pod event', async () => {
+      await browser.get(`${appHost}/ns/${testName}/events`);
+      //TODO: timing flake here only works when going through debugger
+      await browser.wait(until.presenceOf($('.co-sysevent-stream__status')), 10000);
```
Comment (zherman0) [inline frontend/integration-tests/tests/crud.scenario.ts]: @spadgett @benjaminapetersen - Hey guys, this testcase is failing but when I use the debugger it passes.  Any recommendations or thoughts on a better test?

File: frontend/integration-tests/tests/crud.scenario.ts [hunk 7]
```diff
@@ -263,6 +263,55 @@ describe('Kubernetes resource CRUD operations', () => {
...
+      await browser.wait(until.presenceOf($('.co-sysevent__box')), 20000);
+      const sysevent = $$('.co-resource-item__resource-name').first();
+      const myel = sysevent.element(by.linkText(name));
+      await myel.click();
+      await browser.wait(until.urlContains(`/${name}`), 10000);
+      debugger;
```
Comment (benjaminapetersen) [inline frontend/integration-tests/tests/crud.scenario.ts]: We will want the `debugger` statement removed.

**Comment by zherman0** [comment]
@spadgett @benjaminapetersen  - My testcases now work.  I created a new file for them because I plan to add a more complex test in 4.5. This is ready for final review. Thanks.

File: frontend/integration-tests/tests/event.scenario.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
+import { browser, $, element, by, ExpectedConditions as until } from 'protractor';
+import { appHost, checkLogs, checkErrors, testName } from '../protractor.conf';
+import { safeLoad, safeDump } from 'js-yaml';
+import * as _ from 'lodash';
+import * as crudView from '../views/crud.view';
+import * as yamlView from '../views/yaml.view';
+
+describe('Events', () => {
+  const leakedResources = new Set<string>();
```
Comment (benjaminapetersen) [inline frontend/integration-tests/tests/event.scenario.ts]: `leakedResources`?

File: frontend/integration-tests/tests/event.scenario.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    it('displays Events all resources', async () => {
+      await browser.get(`${appHost}/k8s/all-namespaces/events`);
+      await browser.wait(until.presenceOf($('.co-sysevent__box')));
+      expect($('.co-sysevent__box').isPresent()).toBe(true);
+    });
+
+    it('creates a new pod resource instance for event testing', async () => {
+      await browser.get(`${appHost}/k8s/ns/${testName}/pods`);
+      await crudView.isLoaded();
+      await crudView.createYAMLButton.click();
```
Comment (benjaminapetersen) [inline frontend/integration-tests/tests/event.scenario.ts]: Why is the `creatYAMLButton.click()` needed?

Comment (zherman0) [reply frontend/integration-tests/tests/event.scenario.ts]: That's what takes me into the yaml editor, which I use to create the pod.

File: frontend/integration-tests/tests/event.scenario.ts (hunk 8 above)
Comment (zherman0) [reply frontend/integration-tests/tests/event.scenario.ts]: Copy and pasting from another test. Cleaning it up.

File: frontend/integration-tests/tests/event.scenario.ts [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+      await yamlView.saveButton.click();
+      expect(crudView.errorMessage.isPresent()).toBe(false);
+    });
+
+    it('event view displays created pod', async () => {
+      await browser.get(`${appHost}/ns/${testName}/events`);
+      await browser.wait(until.presenceOf($('.co-sysevent-stream__status')), 10000);
+      await browser.wait(until.presenceOf($('.co-sysevent__box')), 10000);
+      await browser.wait(until.presenceOf(element(by.linkText(name))), 10000);
+      await browser.wait(until.presenceOf($(`[data-test-id=${name}]`)));
```
Comment (spadgett) [inline frontend/integration-tests/tests/event.scenario.ts]: Do we need to wait for each of these elements? I'd think we could have only this last wait.

In most cases, we can just use the default timeout. It's unusual you would need something different.

File: frontend/integration-tests/tests/event.scenario.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  });
+
+  describe('Events', () => {
+    const name = `${testName}-pod`;
+    const testLabel = 'automatedTestName';
+
+    it('displays Events all resources', async () => {
+      await browser.get(`${appHost}/k8s/all-namespaces/events`);
+      await browser.wait(until.presenceOf($('.co-sysevent__box')));
+      expect($('.co-sysevent__box').isPresent()).toBe(true);
```
Comment (spadgett) [inline frontend/integration-tests/tests/event.scenario.ts]: This should never evaluate to false since we're waiting for it just above.

I'm not sure what value we get from this test, particularly because we check for a specific pod event below. I would probably just remove this in favor of the pod test.

File: frontend/integration-tests/tests/event.scenario.ts [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    });
+
+    it('creates a new pod resource instance for event testing', async () => {
```
Comment (spadgett) [inline frontend/integration-tests/tests/event.scenario.ts]: It would be simpler to use `kubectl` to create a pod in `beforeAll` and use `kubectl` to delete the pod in `afterAll`. The CRUD tests already cover creating and deleting resources in console really well, so we should keep it simple here. That would avoid duplicating the logic and possible flakes.

**Comment by spadgett** [comment]
Thanks for the tests @zherman0. This is looking a lot simpler 👍

File: frontend/integration-tests/tests/event.scenario.ts (hunk 12 above)
Comment (zherman0) [reply frontend/integration-tests/tests/event.scenario.ts]: Oh so much simpler - I was not aware of execSync - the testcases are so much more straight forward now only checking if the events show up for all and for specific namespaces instead of doing crud stuff.

File: frontend/integration-tests/tests/event.scenario.ts (hunk 11 above)
Comment (zherman0) [reply frontend/integration-tests/tests/event.scenario.ts]: Since we added the route that test the 'all-namespaces" path, I am fine removing this.

Comment (spadgett) [reply frontend/integration-tests/tests/event.scenario.ts]: OK I see the difference now. I'm OK leaving it, although the wait-for and expect check on the same element is odd. It would probably be better to use the expect to check for something else.

File: frontend/integration-tests/tests/event.scenario.ts [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+  beforeAll(() => {
+    execSync(`echo '${JSON.stringify(testpod)}' | kubectl create -n ${testName} -f -`);
+  });
+  afterEach(() => {
+    checkLogs();
+    checkErrors();
+  });
+  afterAll(() => {
+    try {
+      execSync(`kubectl delete --cascade pods ${name} -n ${testName}`);
```
Comment (spadgett) [inline frontend/integration-tests/tests/event.scenario.ts]: cascade not necessary for pods

File: frontend/integration-tests/tests/event.scenario.ts (hunk 11 above)
Comment (spadgett) [reply frontend/integration-tests/tests/event.scenario.ts]: Let's leave it out for now. We can always come back and fill in these tests. Really we should be testing the filters, but I want to make sure the bug fix makes it in.
